pdns/auth-domaincache.hh

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+#include <string>
+#include <map>
+#include "dns.hh"
+#include <boost/version.hpp>
+#include "namespaces.hh"
+using namespace ::boost::multi_index;
+
+#include <boost/multi_index/hashed_index.hpp> 
+
+#include "dns.hh"
+#include "dnspacket.hh"
+#include "lock.hh"
+
+class AuthDomainCache : public boost::noncopyable
+{
+public:
+  AuthDomainCache(size_t mapsCount=1024);
+  ~AuthDomainCache();
+
+  void replace(const vector<tuple<DNSName, int>> &domains);
+
+  bool getEntry(const DNSName &domain, int &backendIndex);
+
+  size_t size() { return *d_statnumentries; } //!< number of entries in the cache
+
+  void setTTL(uint32_t ttl) {
+    d_ttl = ttl;
+  }
+
+  bool isEnabled();
+  bool isExpired();
+private:
+
+  struct CacheEntry
+  {
+    DNSName domain;
+    int backendIndex{-1};
+  };
+
+  struct HashTag{};
+  struct NameTag{};
+  struct SequencedTag{};
+  typedef multi_index_container<
+    CacheEntry,
+    indexed_by <
+      hashed_unique<tag<HashTag>, composite_key<CacheEntry,
+                                                         member<CacheEntry,DNSName,&CacheEntry::domain> > > ,
+      ordered_non_unique<tag<NameTag>, member<CacheEntry,DNSName,&CacheEntry::domain>, CanonDNSNameCompare >,
```

Comment:
I'm not sure we actually need this one? At a quick glance it looks like this cache could be an `unordered_map` rather than a `multi-index` :-)

Reply:
Right. Added as a todo.

Reply:
done, I think?